crates/loans/src/lib.rs

```diff
@@ -0,0 +1,2049 @@
+pub use types::{BorrowSnapshot, EarnedSnapshot, Market, MarketState, RewardMarketState};
+pub use weights::WeightInfo;
+
+mod benchmarking;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+mod farming;
+mod interest;
+#[cfg(test)]
+mod lend_token;
+mod rate_model;
+mod types;
+
+pub mod weights;
+
+type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+type AssetIdOf<T> = <<T as Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::AssetId;
+type BalanceOf<T> = <<T as Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+pub struct OnSlashHook<T>(marker::PhantomData<T>);
+impl<T: Config> OnSlash<T::AccountId, AssetIdOf<T>, BalanceOf<T>> for OnSlashHook<T> {
+    fn on_slash(currency_id: AssetIdOf<T>, account_id: &T::AccountId, amount: BalanceOf<T>) {
+        if currency_id.is_lend_token() {
+            let f = || -> DispatchResult {
+                let underlying_id = Pallet::<T>::underlying_id(currency_id)?;
+                Pallet::<T>::update_reward_supply_index(underlying_id)?;
+                Pallet::<T>::distribute_supplier_reward(underlying_id, account_id)?;
+                Ok(())
+            };
+            if let Err(e) = f() {
+                log::trace!(
+                    target: "loans::on_slash",
+                    "error: {:?}, currency_id: {:?}, account_id: {:?}, amount: {:?}",
+                    e,
+                    currency_id,
+                    account_id,
+                    amount,
+                );
+            }
+        }
+    }
+}
+
+pub struct OnDepositHook<T>(marker::PhantomData<T>);
+impl<T: Config> OnDeposit<T::AccountId, AssetIdOf<T>, BalanceOf<T>> for OnDepositHook<T> {
+    fn on_deposit(currency_id: AssetIdOf<T>, account_id: &T::AccountId, _: BalanceOf<T>) -> DispatchResult {
+        if currency_id.is_lend_token() {
+            let underlying_id = Pallet::<T>::underlying_id(currency_id)?;
+            Pallet::<T>::update_reward_supply_index(underlying_id)?;
+            Pallet::<T>::distribute_supplier_reward(underlying_id, account_id)?;
+        }
+        Ok(())
+    }
+}
+
+pub struct OnTransferHook<T>(marker::PhantomData<T>);
+impl<T: Config> OnTransfer<T::AccountId, AssetIdOf<T>, BalanceOf<T>> for OnTransferHook<T> {
+    fn on_transfer(
+        currency_id: AssetIdOf<T>,
+        from: &T::AccountId,
+        to: &T::AccountId,
+        _: BalanceOf<T>,
+    ) -> DispatchResult {
+        if currency_id.is_lend_token() {
+            let underlying_id = Pallet::<T>::underlying_id(currency_id)?;
+            Pallet::<T>::update_reward_supply_index(underlying_id)?;
+            Pallet::<T>::distribute_supplier_reward(underlying_id, from)?;
+            Pallet::<T>::distribute_supplier_reward(underlying_id, to)?;
+        }
+        Ok(())
+    }
+}
+
+/// Utility type for managing upgrades/migrations.
+#[derive(Encode, Decode, Clone, Copy, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub enum Versions {
```

Comment:
Should we remove these storage versions since we're deploying this pallet for the first time?

Reply:
The default was set to `2` so I'll reset